mlflow/utils/process.py

```diff
@@ -44,8 +45,13 @@ def _exec_cmd(
                       If this argument is specified, `kwargs` cannot contain `env`.
     :param: capture_output: If True, stdout and stderr will be captured and included in an exception
                             message on failure; if False, these streams won't be captured.
-    :param kwargs: Keyword arguments (except `check` and `text`) passed to `subprocess.run`.
-    :return: A `subprocess.CompletedProcess` instance.
+    :param: synchronous: If True, wait process complete and return a CompletedProcess instance,
```

Comment:
I'm not sure if this is a good idea, can we provide two methods (`_exec_cmd_sync` and `_exec_cmd_async`) to simplify the behavior?

Reply:
@harupy Why isn't this a good idea? We do it elsewhere.

Reply:
I just don't prefer a function that returns the union of multiple objects (`Union[obj1, obj2, ...]`). We currently don't use a type checker like mypy, but if we did, such a function might cause an issue like below:

```python
import typing as t


def get_str_or_int(flag: bool) -> t.Union[str, int]:
    return "foo" if flag else 0


r = get_str_or_int(True)
# this works fine but mypy complains `Item "int" of "Union[str, int]" has no attribute "upper"`
r.upper()

# this also works and mypy is happy now but every time we want to call string methods on `r`,
# we need `isinstance`
if isinstance(r, str):
    r.upper()
```

> We do it elsewhere.

True, let's keep the current code.

Reply:
Can we only `Popen` instead of using both `Popen` and `subprocess.run`?

```python
import subprocess


def _exec_cmd(
    cmd,
    *,
    throw_on_error=True,
    extra_env=None,
    capture_output=True,
    synchronous=True,
    **kwargs,
):
    ...
    if capture_output:
        kwargs["stdout"] = subprocess.PIPE
        kwargs["stderr"] = subprocess.PIPE

    process = subprocess.Popen(
        cmd,
        env=env,
        text=True,
        **kwargs,
    )
    if not synchronous:
        return process

    stdout, stderr = process.communicate()
    returncode = process.poll()
    comp_process = subprocess.CompletedProcess(
        process.args,
        returncode=returncode,
        stdout=stdout,
        stderr=stderr,
    )
    if throw_on_error and returncode != 0:
        raise ShellCommandException.from_completed_process(comp_process)
    return comp_process

```